osbuild/monitor.py

```diff
@@ -9,17 +9,175 @@
 """
 
 import abc
+import copy
 import datetime
+import hashlib
 import json
 import os
 import sys
 import time
-from typing import Dict
+from typing import Dict, Optional, Set
 
 import osbuild
 from osbuild.util.term import fmt as vt
 
 
+def omitempty(d: dict):
+    """ Omit None and empty string ("") values from the given dict """
+    for k, v in list(d.items()):
+        if v is None or v == "":
+            del d[k]
+        elif isinstance(v, dict):
+            omitempty(v)
+    return d
+
+
+class Context:
+    """Context for a single log entry. Automatically calculates hash/id when read."""
+
+    def __init__(self,
+                 origin: Optional[str] = None,
+                 pipeline: Optional[osbuild.Pipeline] = None,
+                 stage: Optional[osbuild.Stage] = None):
+        self._origin = origin
+        self._pipeline_name = pipeline.name if pipeline else None
+        self._pipeline_id = pipeline.id if pipeline else None
+        self._stage_name = stage.name if stage else None
+        self._stage_id = stage.id if stage else None
+        self._id = None
+        self._id_history: Set[str] = set()
+
+    def __setattr__(self, name, value):
+        super().__setattr__(name, value)
+        # reset "_id" on any write so that the hash is automatically recalculated
+        if name != "_id":
+            super().__setattr__("_id", None)
+
+    def with_origin(self, origin: Optional[str]) -> "Context":
+        """
+        Return a Context with the given origin but otherwise identical.
+
+        Note that if the origin is empty or same it will return self.
+        """
+        if origin is None or origin == self._origin:
+            return self
+        ctx = copy.copy(self)
+        ctx.set_origin(origin)
+        return ctx
+
+    @property
+    def origin(self):
+        return self._origin
+
+    def set_origin(self, origin: str):
+        self._origin = origin
+
+    @property
+    def pipeline_name(self):
+        return self._pipeline_name
+
+    @property
+    def pipeline_id(self):
+        return self._pipeline_id
+
+    def set_pipeline(self, pipeline: osbuild.Pipeline):
+        self._pipeline_name = pipeline.name
+        self._pipeline_id = pipeline.id
+
+    @property
+    def stage_name(self):
+        return self._stage_name
+
+    @property
+    def stage_id(self):
+        return self._stage_id
+
+    def set_stage(self, stage: osbuild.Stage):
+        self._stage_name = stage.name
+        self._stage_id = stage.id
+
+    @property
+    def id(self):
+        if self._id is None:
+            self._id = hashlib.sha256(json.dumps(self._dict()).encode()).hexdigest()
```

Comment:
Since we officially support Python 3.6 and dictionaries keeping insertion orders was technically an implementation detail there we'd need explicit recursive sorting.

Passing `sort_keys=True` to `json.dumps` should be sufficient.

Reply:
Nice catch! Thanks a lot!